src/transformers/models/clip/modeling_tf_clip.py

```diff
@@ -0,0 +1,1528 @@
+
+        batch_size, num_channels, height, width = shape_list(pixel_values)
+
+        # When running on CPU, `tf.nn.conv2d` doesn't support `NCHW` format.
+        # So change the input format from `NCHW` to `NHWC`.
+        # shape = (batch_size, in_height, in_width, in_channels=num_channels)
+        pixel_values = tf.transpose(pixel_values, perm=(0, 2, 3, 1))
+
+        patch_embeds = self.patch_embedding(pixel_values)
+
+        # Change the 2D spatial dimensions to a single temporal dimension.
+        # shape = (batch_size, num_patches, out_channels=embed_dim)
+        patch_embeds = tf.reshape(tensor=patch_embeds, shape=(batch_size, self.num_patches, -1))
+
+        # add the [CLS] token to the embedded patch tokens
+        class_embeds = tf.broadcast_to(self.class_embedding, shape=(batch_size, 1, self.embed_dim))
+        embeddings = tf.concat((class_embeds, patch_embeds), axis=1)
+
+        embeddings = embeddings + self.position_embedding
+
+        return embeddings
+
+
+class TFCLIPTextEmbeddings(tf.keras.layers.Layer):
+    def __init__(self, config: CLIPTextConfig, **kwargs):
+        super().__init__(**kwargs)
+
+        self.embed_dim = config.hidden_size
+        self.vocab_size = config.vocab_size
+
+        self.config = config
+
+    def build(self, input_shape: tf.TensorShape):
+
+        with tf.name_scope("token_embedding"):
+            self.weight = self.add_weight(
+                shape=(self.vocab_size, self.embed_dim),
+                initializer=get_initializer(self.config.initializer_range),
+                trainable=True,
+                name="weight",
+            )
+
+        with tf.name_scope("position_embedding"):
+            self.position_embedding = self.add_weight(
+                shape=(self.config.max_position_embeddings, self.embed_dim),
+                initializer=get_initializer(self.config.initializer_range),
+                trainable=True,
+                name="embeddings",
+            )
+
+        super().build(input_shape)
+
+    def call(
+        self,
+        input_ids: tf.Tensor = None,
+        position_ids: tf.Tensor = None,
+        inputs_embeds: tf.Tensor = None,
+    ) -> tf.Tensor:
+        """
+        Applies embedding based on inputs tensor.
+
+        Returns:
+            final_embeddings (:obj:`tf.Tensor`): output embedding tensor.
+        """
+        if input_ids is None and inputs_embeds is None:
+            raise ValueError("You have to specify either input_ids or inputs_embeds")
+
+        if inputs_embeds is None:
+            inputs_embeds = tf.gather(params=self.weight, indices=input_ids)
+
+        input_shape = shape_list(inputs_embeds)[:-1]
+
+        if position_ids is None:
+            position_ids = tf.expand_dims(tf.range(start=0, limit=input_shape[-1]), axis=0)
+
+        position_embeds = tf.gather(params=self.position_embedding, indices=position_ids)
+        position_embeds = tf.tile(input=position_embeds, multiples=(input_shape[0], 1, 1))
```

Comment:
@Rocketknight1 Good point. Yes, it is possible. However

- The line `position_embeds = tf.tile(input=position_embeds, multiples=(input_shape[0], 1, 1))` is everywhere in TF model scripts. If we change here, probably it is better to change all (?)
- `tf.broadcast_to` is not perfect as mentioned in TF doc [broadcast_to](https://www.tensorflow.org/api_docs/python/tf/broadcast_to)

  ```
  When doing broadcasted operations such as multiplying a tensor by a scalar,
  broadcasting (usually) confers some time or space benefit, as the broadcasted
  tensor is never materialized.
  
  However, broadcast_to does not carry with it any such benefits. The newly-created
  tensor takes the full memory of the broadcasted shape. (In a graph context, broadcast_to
  might be fused to subsequent operation and then be optimized away, however.)
  ```

But after diving deeper, I found something more severe: 

The `...INPUTS_DOCSTRING` (everywhere in transformers) specifies `position_ids` should have shape `(batch_size, sequence_length)`. If a user specify `position_ids` manually with the full `batch_size` (very unlikely though), then the line `tf.tile` shouldn't be run! (otherwise the `batch_size` will become squared). 

If we use `tf.broadcast_to`, we don't have such issue.

Could you discuss with @sgugger  and @LysandreJik  to see what to do about this, please?

Reply:
I don't think we need `tf.broadcast_to` either - because the first dimension of `position_embeds` is 1, we can just remove the tile line entirely, and position_embeds will still be correctly broadcast in the next line. In other words, the code would just be:
```
position_embeds = tf.gather(params=self.position_embedding, indices=position_ids)
final_embeddings = inputs_embeds + position_embeds
```

I did some quick testing and this seems to work fine. However, it's just a nit, and the performance impact will likely be small anyway, so don't worry about it too much! We can also just leave it, and I'll make a note later to check the codebase for instances of `tile` that should be replaced with implicit broadcasting.

Reply:
Also, thinking about it more, using broadcasting resolves your issue with the users specifying a full batch of `position_ids`, so maybe it's more than just a minor performance nit!

Reply:
> 

Thanks, @Rocketknight1 

Well, implicit broadcasting will still work when users specifying a full batch - there won't be any broadcasting in this case.

But as you said, we can leave it to another PR if necessary :)